akka-stream/src/main/scala/akka/stream/remote/scaladsl/SinkRef.scala

```diff
@@ -0,0 +1,228 @@
+    ???
+  }
+}
+
+///**
+// * INTERNAL API
+// */
+//@InternalApi private[akka] final class SinkRefModule[In](val attributes: Attributes, shape: SinkShape[In]) extends SinkModule[In, ActorRef](shape) {
+//
+//  override def create(context: MaterializationContext): (Subscriber[In], ActorRef) = {
+//    val system = ActorMaterializerHelper.downcast(context.materializer)
+//    val subscriberRef = system.actorOf(context.copy(islandName = ""), SinkRefActor.props)
+//    (akka.stream.actor.ActorSubscriber[In](subscriberRef), subscriberRef)
+//  }
+//
+//  override protected def newInstance(shape: SinkShape[In]): SinkModule[In, ActorRef] = new SinkRefModule[In](attributes, shape)
+//  override def withAttributes(attr: Attributes): SinkModule[In, ActorRef] = new SinkRefModule[In](attr, amendShape(attr))
+//}
+
+/**
+ * This stage can only handle a single "sender" (it does not merge values);
+ * The first that pushes is assumed the one we are to trust
+ */
+final class SinkRefTargetSource[T] extends GraphStageWithMaterializedValue[SourceShape[T], Future[SinkRef[T]]] {
+  val out: Outlet[T] = Outlet[T]("SinkRef.source.out")
+  override def shape = SourceShape.of(out)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes) = {
+    val promise = Promise[SinkRef[T]]()
+
+    val logic = new GraphStageLogic(shape) with StageLogging with OutHandler {
+      private[this] lazy val streamRefsMaster = StreamRefsMaster(ActorMaterializerHelper.downcast(materializer).system)
+
+      private[this] var self: GraphStageLogic.StageActor = _
+      private[this] lazy val selfActorName = streamRefsMaster.nextSinkRefTargetSourceName()
+      private[this] implicit def selfSender: ActorRef = self.ref
+
+      val initialDemand = 4L // TODO get from config as well as attributes
+
+      var localCumulativeDemand = 0L
+      var remotePartner: ActorRef = _
+
+      override def preStart(): Unit = {
+        self = getStageActor(initialReceive, name = selfActorName)
+
+        //        receiverRef = master.allocateReceiver(self.ref)
+        log.warning("Allocated receiver: {}", self.ref)
+
+        promise.success(new SinkRef(self.ref, initialDemand))
+      }
+
+      override def onPull(): Unit = {
+        localCumulativeDemand += 1
+        triggerDemand()
+      }
+
+      // FIXME this should have some smarter strategy than pulling one by one
+      def triggerDemand(): Unit = if (remotePartner ne null) {
+        localCumulativeDemand += 1
+        remotePartner ! StreamRefs.CumulativeDemand(localCumulativeDemand)
+        log.debug("[{}] Demanding until {}", selfActorName, StreamRefs.CumulativeDemand(localCumulativeDemand))
+      }
+
+      def runningReceive(activeSender: ActorRef): ((ActorRef, Any)) ⇒ Unit = {
+        case (`activeSender`, StreamRefs.SequencedOnNext(seqNr, payload: T)) ⇒
+          // FIXME fail or wait if the sequence number is not strictly in order with the expected one
```

Comment:
if seqNr is not the expected one we should fail, since we will not implement resending of elements

Reply:
Yeah, commented elsewhere on this that this will be fail only, no wait, but seems comment got lostin github ui